### PR TITLE
Optimize SimpleDatabase#bulkLookup

### PR DESCRIPTION
**Pull request by Glease**
See https://github.com/Glease/SimpleDatabaseBenchmark for related benchmark

Signed-off-by: Glease <4586901+Glease@users.noreply.github.com>

### CODE REVIEW

**Comment by Skiftcha** [comment]
good work done
maybe we can completely remove merge join method and simply replace TreeMap with HashMap to make naive search faster?
image
image

**Comment by Dream-Master** [comment]
@Skiftcha can i merge this or did it need more reviews ?

File: src/main/java/betterquesting/api2/storage/LookupLogicType.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package betterquesting.api2.storage;
+
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static betterquesting.api2.storage.SimpleDatabase.*;
+
+enum LookupLogicType
+{
+	ArrayCache(db -> db.mapDB.size() < CACHE_MAX_SIZE && db.mapDB.size() < SPARSE_RATIO * (db.mapDB.lastKey() - db.mapDB.firstKey()), ArrayCacheLookupLogic::new),
```
Comment (Skiftcha) [inline src/main/java/betterquesting/api2/storage/LookupLogicType.java:10]: db.mapDB.size() > SPARSE_RATIO * (db.mapDB.lastKey() - db.mapDB.firstKey())

File: src/main/java/betterquesting/api2/storage/SimpleDatabase.java [hunk 2]
```diff
@@ -1,21 +1,66 @@
...
-    
+    private LookupLogicType type = null;
+    private LookupLogic<T> logic = null;
+
+    private LookupLogic<T> getLookupLogic()
+    {
+        if(type == null)
+        {
+            LookupLogicType newType = LookupLogicType.determine(this);
+            if(newType != type)
```
Comment (Skiftcha) [inline src/main/java/betterquesting/api2/storage/SimpleDatabase.java:34]: no need to check this

File: src/main/java/betterquesting/api2/storage/LookupLogicType.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	}
+
+	static LookupLogicType determine(SimpleDatabase<?> db)
+	{
+		for(LookupLogicType type : values())
+		{
+			if(type.shouldUse.test(db))
+				return type;
+		}
+		return values()[values().length - 1];
```
Comment (Skiftcha) [inline src/main/java/betterquesting/api2/storage/LookupLogicType.java:29]: return Naive;

**Review by Skiftcha — CHANGES_REQUESTED** [review]
and still remains question about need of SortMerge and TreeMap usage instead of HashMap

File: src/main/java/betterquesting/api2/storage/SimpleDatabase.java (hunk 2 above)
Comment (Glease) [reply src/main/java/betterquesting/api2/storage/SimpleDatabase.java:34]: if you don't check this you rebuild the cache on every `bulkLookup` invocation. not wise.

File: src/main/java/betterquesting/api2/storage/LookupLogicType.java (hunk 1 above)
Comment (Glease) [reply src/main/java/betterquesting/api2/storage/LookupLogicType.java:10]: good catch. how could I miss this

**Comment by Glease** [comment]
I'm currently talking with Funwayguy in DM and is looking to optimize this a bit more (so please don't merge as of now). For the HashMap thing his opinion is as follow

> I am also keeping TreeMap over HashMap despite the performance cost due to its pre-sorted nature being far more convenient for display, searches and building caches

I guess we could use it as a replacement of naive lookup so it doesn't break/slow down other stuff.

**Comment by Prometheus0000** [comment]
BTW, with how many quests we have, would it be worthwhile to make each tab it's own file (or something), or is that more effort than it's worth? Especially since you'd have quite a few.

**Comment by Glease** [comment]
Probably too much effort and will require a drastic change to quest storage (not visible to end user) and quest design (visible to end user and require migrating data over)

File: src/main/java/betterquesting/api2/storage/SimpleDatabase.java (hunk 2 above)
Comment (Glease) [reply src/main/java/betterquesting/api2/storage/SimpleDatabase.java:34]: Wait a minute this seems to be code I forget to clean up. You are right here

File: src/main/java/betterquesting/api2/storage/SimpleDatabase.java [hunk 4]
```diff
@@ -1,21 +1,63 @@
...
+
     private final BitSet idMap = new BitSet();
-    private List<DBEntry<T>> refCache = null;
-    
+    private LookupLogicType type = null;
+    private LookupLogic<T> logic = null;
+
+    private LookupLogic<T> getLookupLogic()
+    {
+        if(type == null)
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/storage/SimpleDatabase.java:31]: invert this if to exit early and reduce nesting level

File: src/main/java/betterquesting/api2/storage/LookupLogic.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public void onDataChange()
+    {
+        refCache = null;
+    }
+
+    public List<DBEntry<T>> getRefCache()
+    {
+        if(refCache == null)
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/storage/LookupLogic.java:25]: invert this if to exit early and reduce nesting level

File: src/main/java/betterquesting/api2/storage/ArrayCacheLookupLogic.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                list.add(element);
+            }
+        }
+        return list;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void computeCache()
+    {
+        if(cache == null)
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/storage/ArrayCacheLookupLogic.java:61]: invert this if to exit early and reduce nesting level

File: src/main/java/betterquesting/api2/storage/ArrayCacheLookupLogic.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        super.onDataChange();
+        cache = null;
+        offset = -1;
+    }
+
+    @Override
+    public List<DBEntry<T>> getRefCache()
+    {
+        if(refCache == null)
```
Comment (bartimaeusnek) [inline src/main/java/betterquesting/api2/storage/ArrayCacheLookupLogic.java:27]: invert this if to exit early and reduce nesting level

**Review by bartimaeusnek — COMMENTED** [review]
Just some Code-Style stuff ~~~

**Comment by Glease** [comment]
Changed naive lookup to use a hashmap instead of the default treemap.

So the PR as it is now should work fine enough for our pack. The other optimization I was talking about with Funwayguy will probably show up later in another PR.

File: src/main/java/betterquesting/api2/storage/LookupLogicType.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static betterquesting.api2.storage.SimpleDatabase.*;
+
+enum LookupLogicType
+{
+	ArrayCache(db -> db.mapDB.size() < CACHE_MAX_SIZE && db.mapDB.size() > SPARSE_RATIO * (db.mapDB.lastKey() - db.mapDB.firstKey()), ArrayCacheLookupLogic::new),
+	SortMergeJoin(db -> db.mapDB.size() > SORT_MERGE_JOIN_THRESHOLD, SortMergeJoinLookupLogic::new),
```
Comment (Skiftcha) [inline src/main/java/betterquesting/api2/storage/LookupLogicType.java:11]: do we need SortMergeJoin now? HashMap was always better in my benchmark
but i didnt test with TIntObjectMap yet (i think should be the same or maybe a bit faster)

Comment (Glease) [reply src/main/java/betterquesting/api2/storage/LookupLogicType.java]: Yeah both HashMap and TIntObjectHashMap works faster than sort merge join in any situation. I removed it in the latest commit.

**Review by Skiftcha — APPROVED** [review]

**Pull request merged**